### PR TITLE
Add new option: DiscoveryIgnoreReplicaHostnameFilters

### PR DESCRIPTION
**Pull request by sjmudd**
This is used to allow us to avoid false positive discoveries of servers which look like replicas but which are not. It also allows us to not try to discover servers which we may not be able to reach
due to firewall holes even though we can see them referenced by a master.

- [x] contributed code is using same conventions as original code
- [x] code is formatted via `gofmt` (please avoid `goimports`)
- [x] code is built via `./build.sh`
- [x] code is tested via `go test ./go/...`

### CODE REVIEW

File: go/inst/instance_dao.go [hunk 1]
```diff
@@ -198,6 +198,16 @@ func (instance *Instance) checkMaxScale(db *sql.DB, latency *stopwatch.NamedStop
...
+// RegexpMatchPattern returns true if s matches any of the provided regexpPatterns
+func RegexpMatchPatterns(s string, regexpPatterns []string) bool {
+	for _, filter := range regexpPatterns {
+		if matched, err := regexp.MatchString(filter, s); err == nil && matched {
+			return true
+		}
+	}
+	return false
+}
+
```
Comment (shlomi-noach) [inline go/inst/instance_dao.go]: please move this function somewhere more generic, such as `go/inst/instance_utils.go`

File: go/inst/instance_dao.go [hunk 2]
```diff
@@ -1257,22 +1265,16 @@ func ReadClusterCandidateInstances(clusterName string) ([](*Instance), error) {
...
 		}
-
 		if !instance.IsLastCheckValid {
-			skipThisHost = true
-		}
-		if !skipThisHost {
-			result = append(result, instance)
+			continue
 		}
+		result = append(result, instance)
```
Comment (shlomi-noach) [inline go/inst/instance_dao.go:1267]: :+1:

File: go/logic/orchestrator.go [hunk 3]
```diff
@@ -216,7 +216,13 @@ func discoverInstance(instanceKey inst.InstanceKey) {
...
 	// Investigate replicas:
 	for _, replicaKey := range instance.SlaveHosts.GetInstanceKeys() {
-		replicaKey := replicaKey
+		replicaKey := replicaKey // not needed? no concurrency here?
+
+		// Avoid noticing some hosts we would otherwise discover
+		if inst.RegexpMatchPatterns(replicaKey.Hostname, config.Config.DiscoveryIgnoreReplicaHostnameFilters) {
+			continue
+		}
+
```
Comment (shlomi-noach) [inline go/logic/orchestrator.go:233]: :+1:

**Review by shlomi-noach — CHANGES_REQUESTED** [review]

File: go/inst/instance_dao.go (hunk 1 above)
Comment (sjmudd) [reply go/inst/instance_dao.go]: ok. will change.

**Comment by sjmudd** [comment]
I relocated the function as requested.

File: go/inst/instance_dao2_test.go [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+		{"hostname", []string{"ho.tname2"}, false},
+		{"hostname", []string{"ho.*me"}, true},
+	}
+
+	for _, p := range patterns {
+		if match := RegexpMatchPatterns(p.s, p.patterns); match != p.expected {
+			t.Errorf("RegexpMatchPatterns failed with: %q, %+v, got: %+v, expected: %+v", p.s, p.patterns, match, p.expected)
+		}
+	}
+}
```
Comment (shlomi-noach) [inline go/inst/instance_dao2_test.go]: Sorry, this should move to be in `instance_utils_test.go`, following the move of `RegexpMatchPatterns` to `instance_utils.go`

**Review by shlomi-noach — CHANGES_REQUESTED** [review]

**Comment by sjmudd** [comment]
Thanks for your patience, and hopefully this works for you.
Also fixed a comment typo to keep golint happy.

File: go/logic/orchestrator.go [hunk 5]
```diff
@@ -224,7 +224,13 @@ func discoverInstance(instanceKey inst.InstanceKey) {
 
 	// Investigate replicas:
 	for _, replicaKey := range instance.SlaveHosts.GetInstanceKeys() {
-		replicaKey := replicaKey
+		replicaKey := replicaKey // not needed? no concurrency here?
```
Comment (shlomi-noach) [inline go/logic/orchestrator.go:227]: correct, no concurrency here; keep it anyhow...

**Review by shlomi-noach — APPROVED** [review]

**Pull request merged**